### PR TITLE
Document `doc(keyword)` unstable attribute

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @camelid

### CODE REVIEW

File: src/doc/rustdoc/src/unstable-features.md [hunk 1]
```diff
@@ -135,8 +135,20 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
 
 Since primitive types are defined in the compiler, there's no place to attach documentation
-attributes. The `#[doc(primitive)]` attribute is used by the standard library to provide a way to generate
-documentation for primitive types, and requires `#![feature(doc_primitive)]` to enable.
+attributes. The `#[doc(primitive)]` attribute is used by the standard library to provide a way
+to generate documentation for primitive types, and requires `#![feature(doc_primitive)]` to enable.
+
+## Document keywords
+
+Keywords are documented in the standard library (look for "match" for example). However,
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:143]: Just to make it extra clear:

```suggestion
Rust keywords are documented in the standard library (look for `match` for example). However,
```

File: src/doc/rustdoc/src/unstable-features.md [hunk 2]
```diff
@@ -135,8 +135,20 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
+you might write a macro or a proc-macro adding new keywords. If you want to document them,
+you can by using `#[doc(keyword)]` on an empty module. Example:
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:145]: Do we really want to be recommending that macro authors use `#[doc(keyword)]`? My understanding was that both `#[doc(primitive)]` and `#[doc(keyword)]` are supposed to be only used by the stdlib and are perma-unstable.

File: src/doc/rustdoc/src/unstable-features.md [hunk 3]
```diff
@@ -135,8 +135,20 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
+
+```rust,ignore (needs nightly feature)
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:147]: Why is this ignored? Doctests are run on nightly Rust. Just add `#![feature(doc_keyword)]`.

**Review by camelid — CHANGES_REQUESTED** [review]
I don't think we should be documenting `#[doc(keyword)]` and `#[doc(primitive)]` beyond saying, "They're permanently unstable and only meant to be used by the standard library."

**Comment by camelid** [comment]
cc @rust-lang/rustdoc for discussion about whether we should be documenting `#[doc(keyword)]`, and if so, to what extent.

**Comment by GuillaumeGomez** [comment]
> I don't think we should be documenting `#[doc(keyword)]` and `#[doc(primitive)]` beyond saying, "They're permanently unstable and only meant to be used by the standard library."

Except that if you use macro and proc-macro, you might want to add new keywords. Wouldn't it make sense to document them then? I still think that this feature should get stabilized at some point.

File: src/doc/rustdoc/src/unstable-features.md (hunk 2 above)
Comment (Manishearth) [reply src/doc/rustdoc/src/unstable-features.md:145]: Yeah, I don't think this feature should be used by others for that purpose, they should document it in the macro itself.

**Comment by camelid** [comment]
> > I don't think we should be documenting `#[doc(keyword)]` and `#[doc(primitive)]` beyond saying, "They're permanently unstable and only meant to be used by the standard library."
> 
> Except that if you use macro and proc-macro, you might want to add new keywords. Wouldn't it make sense to document them then? I still think that this feature should get stabilized at some point.

The "keywords" in macros that you are describing are not really keywords; they are limited to the scope of the macro invocation. Outside of an invocation of that macro, they are not keywords. As @Manishearth [said](https://github.com/rust-lang/rust/pull/90398#discussion_r739516354), custom macro keywords should be documented in the macro itself, not using `#[doc(keyword)]`.

**Comment by camelid** [comment]
In some ways, I wonder if it would make sense for us to put `#[doc(primitive)]` and `#[doc(keyword)]` behind the same feature gate, named something like `rustdoc_internals`. That would underscore that the attributes are more like implementation details.

**Comment by Manishearth** [comment]
Right, macros can't actually add new keywords, they can in their own context parse identifiers in a way that may seem keywordlike. Which can be documented on the macro itself.

**Comment by GuillaumeGomez** [comment]
That makes sense. I'll reword the documentation then!

@camelid Putting them behind the same feature would make sense. I'll send a PR for that. :)

File: src/doc/rustdoc/src/unstable-features.md (hunk 3 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:147]: :+1:

File: src/doc/rustdoc/src/unstable-features.md [hunk 4]
```diff
@@ -135,8 +135,22 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
 
 Since primitive types are defined in the compiler, there's no place to attach documentation
-attributes. The `#[doc(primitive)]` attribute is used by the standard library to provide a way to generate
-documentation for primitive types, and requires `#![feature(doc_primitive)]` to enable.
+attributes. The `#[doc(primitive)]` attribute is used by the standard library to provide a way
+to generate documentation for primitive types, and requires `#![feature(doc_primitive)]` to enable.
+
+## Document keywords
+
+Rust keywords are documented in the standard library (look for "match" for example).
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:143]: I think you missed my earlier request to change this to `match`.

```suggestion
Rust keywords are documented in the standard library (look for `match` for example).
```

File: src/doc/rustdoc/src/unstable-features.md [hunk 5]
```diff
@@ -135,8 +135,22 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
+
+To do so, the `#[doc(keyword = "...")]` attribute is used. Example:
+
+```rust
+#![feature(doc_keyword)]
+
+#[doc(keyword = "Keyword")]
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:150]: nit: Keywords are usually lowercase:

```suggestion
#[doc(keyword = "keyword")]
```

**Review by camelid — COMMENTED** [review]
Two nits, but also: wasn't the plan to change the docs here so they make it clear that these attributes are for internal use only? Or were you going to do that in the `rustdoc_internals` PR?

**Comment by GuillaumeGomez** [comment]
> Two nits, but also: wasn't the plan to change the docs here so they make it clear that these attributes are for internal use only?

We don't know what the future will be made of. Maybe we will allow to create new keywords in the future? Who knows? At least now, we simply explain how we do it in the standard library and that's it.

Fixing the two nits quickly.

**Comment by jyn514** [comment]
@GuillaumeGomez if we change our minds in the future, we can change the documentation at the same time. For now I think we should make it clear that these aren't intended to be stabilized.

**Comment by GuillaumeGomez** [comment]
We didn't say so anywhere else as far as I can see, so seems weird to start it now...

**Comment by jyn514** [comment]
@GuillaumeGomez the whole point of #90420 is that these are not intended to be used by anyone other than the standard library.

**Comment by GuillaumeGomez** [comment]
Fine, fine...

**Comment by GuillaumeGomez** [comment]
Done.

File: src/doc/rustdoc/src/unstable-features.md [hunk 6]
```diff
@@ -135,8 +135,24 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
 
 Since primitive types are defined in the compiler, there's no place to attach documentation
-attributes. The `#[doc(primitive)]` attribute is used by the standard library to provide a way to generate
-documentation for primitive types, and requires `#![feature(doc_primitive)]` to enable.
+attributes. The `#[doc(primitive)]` attribute is used by the standard library to provide a way
+to generate documentation for primitive types, and requires `#![feature(doc_primitive)]` to enable.
+
+## Document keywords
+
+This is for rust compiler internal use only.
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:143]:
```suggestion
This is for Rust compiler internal use only.
```

Comment (camelid) [reply src/doc/rustdoc/src/unstable-features.md:143]: While you're at it, can you add this sentence to the beginning of the "Document primitives" section too?

File: src/doc/rustdoc/src/unstable-features.md [hunk 7]
```diff
@@ -135,8 +135,24 @@ Book][unstable-masked] and [its tracking issue][issue-masked].
...
+Rust keywords are documented in the standard library (look for `match` for example).
+
+To do so, the `#[doc(keyword = "...")]` attribute is used. Example:
+
+```rust
+#![feature(doc_keyword)]
+
+#[doc(keyword = "keyword")]
+/// Some documentation about the keyword.
+mod empty_mod {}
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:154]: nit: Usually the attribute comes immediately before the type's definition:

```suggestion
/// Some documentation about the keyword.
#[doc(keyword = "keyword")]
mod empty_mod {}
```

**Review by camelid — COMMENTED** [review]
r=me with the following three comments addressed

File: src/doc/rustdoc/src/unstable-features.md (hunk 6 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:143]: Sure!

File: src/doc/rustdoc/src/unstable-features.md (hunk 7 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:154]: In the std docs they don't, but I guess it doesn't really matter. I'll update. :)

**Pull request merged**